text/0757-default-modifier-manager.md

```diff
@@ -0,0 +1,276 @@
+
+<details><summary>Example</summary>
+
+<!--
+  If you're reading this RFC non-rendered, ignore the "jsx" language tag.
+  "jsx" is incorrect, and isn't perfect, but "gjs" syntax highlighting
+  hasn't been added to GitHub's highlighting yet.
+-->
+
+This could be an entire component
+
+```jsx
+const draggable = element => {
+  element.addEventListener( /*... */ );
+
+  return () => element.removeEventListener( /* ... */ );
+};
+
+<template>
+  <div {{draggable}}>text content A</div>
+  <div {{draggable}}>text content B</div>
+  <div {{draggable}}>text content C</div>
+  <div {{draggable}}>text content D</div>
+</template>
+```
+_Note: `<template>` is experimental syntax, and should not be be the focus of this example_
+
+
+</details>
+
+
+## Detailed design
+
+_A Default Manager is not something that can be chosen by the user, but is baked in to the framework
+as a default so that a user doesn't have to build something to use a non-framework-specific variant
+of the three constructs: Helpers, Modifiers, and Components._
+
+The desired usage of a plain function in a template should be:
+ - convenient
+ - reduce boilerplate
+ - be easily portable to JS for developers' mental model of how template and JS interact.
+
+Which results in:
+ - default to positional parameters
+ - all named arguments are grouped into an "options object" as the last parameter.
+   this happens to align with the _syntax_ of modifier invocation where named arguments may not appear
+   before the last positional argument.
+
+#### Example with mixed params
+
+```hbs
+<div {{this.myModifier 1 2 op="add"}}></div>
+```
+would be an example invocation of a function with the following signature
+expressed as TypeScript for clarity:
+```ts
+type Options = { op: 'add' | 'subtract' }
+class A {
+  myModifier = (element: Element, first: number, second: number, options: Options) => {
+    // ...
+  }
+}
+```
+
+#### Example with only positional parameters
+
+```hbs
+<div {{this.myModifier 1 2}}></div>
+```
+Because there are no named arguments passed in, the develope may opt to omit the options argument from the method signature.
+```ts
+class A {
+  myModifier = (element: Element, first: number, second: number) => {
+    // ..
+  }
+}
+```
+however, the options argument is still passed, but is equivalent to `{}`.
+
+
+#### Example default modifier implementation
+
+The type signature for any function-modifier could be:
+
+```ts
+import type { CapturedArguments as Arguments } from '@glimmer/interfaces';
+
+type FnArgs<Args extends Arguments = Arguments> =
+  | [Element, ...Args['positional'], Args['named']]
+  | [Element, ...Args['positional'], {}];
+
+type FunctionModifier<
+  Args extends Arguments = Arguments,
+  ModifierArgs extends FnArgs<Args> = FnArgs<Args>
+> =
+  | ((...args: ModifierArgs) => void) // setup, with no teardown
+  | ((...args: ModifierArgs) => () => void) // setup with teardown
+
+class A {
+  myModifierA: FunctionModifier = (element, a, b, c, options) => {
```

Comment:
The reason the aren't actually helpful: there's no value on the TS side to providing a form where you can simply say "any old modifier" but *not* actually specify what its elements are. You end up with basically useless types within the body if you do that.

In general, we (meaning on the Typed Ember/Glint side) also won't recommend people import and use `FunctionModifier`. We will just recommend that they define a function which has the right form, and Glint will let them know if it doesn't pass! But for the sake of supporting this where folks *do* want it, the version I showed above gets the job done.

Reply:
I like that plan!